### DISCUSSION TITLE
Is it possible to run patterns from command line?

### DISCUSSION THREAD

**Discussion by pshirshov** [WerWolv/ImHex#1940] [Pattern Language Q&A] [answered]
It would be awesome if we can run patterns as standalone CLI scripts against files and stdin, that would save us a lot of effort in many cases. Although can't figure out if that's possible. Is it?

**Comment by WerWolv** [comment, marked as answer]
ImHex has a command line interface that you can use. Run `imhex --help` to see what's available. The sub command you're looking for is `imhex --pl`. There's also a standalone cli for the Pattern Language itself, without any ImHex stuff that you can build yourself from the library sources.

**Reply by pshirshov** [reply]
Indeed it works, thank you. 

Although it loves to crash randomly (at least, on ARM Macs): 
<img width="283" alt="image" src="https://github.com/user-attachments/assets/09277e08-1045-4241-bdd5-3c701a972db6">

Like it might literally fail in 50% of the time while being invoked on the same input file. Also, it always finishes parsing and prints the output and then shows this nasty dialog in the end.

Also, it seems like it doesn't support stdin, am I right?

**Reply by msm-code** [reply]
You can read data from stdin just fine:

```
$ echo a | imhex --pl format --pattern f --input /dev/stdin --output tt
Wrote pattern data to tt
```

for some reason this doesn't work with patterns:

```
$ echo "u8 a @ 1;" | imhex --pl format --pattern /dev/stdin --input /etc/passwd --output tt1
fish: Process 1946761, 'imhex' from job 1, 'echo "u8 a @ 1;" | imhex --pl f…' terminated by signal SIGABRT (Abort)
```
